README
A tool to compare 2 Puppet catalogs.

PLEASE NOTE: This is still early code, its a work in progress specifically I am working on 0.24.x compatibility

While upgrading versions of Puppet you want to ensure that
no unexpected changes will be made prior to doing the upgrade.

This tool will allow you to diff catalogs created by different
versions of Puppet.  This will let you guage the impact of a
Puppet upgrade before actually touching any of your nodes.

This tool is delivered as a Puppet Face. It thus requires
a Puppet installation to properly run.

This is the first version of the code that works, it's not
clean code or tested throughly, as I use it to do upgrades in
the near future there will no doubt be refinements.

The diff tool recognizes catalogs in yaml, marshall, or pson format.

Validation Process:

 - Grab a catalog from your existing machine running the old version
 - Configure your new Puppet master, copy the facts from your old master
   to the new one
 - Compile the catalog for this host on the new master:

      puppet master --compile fqdn > fqdn.pson

 - At this point you should have 2 different catalogs. To compare them run:

      puppet catalog diff <catalog1> <catalog2>

Example Output:

During the transition of 0.24.x to 0.25.x there was a serialization bug
that resulted in unexpected file content changes, I've recreated this bug
in a tiny 2 resource catalog, the output below shows how this tool would
have highlighted this bug prior to upgrading any nodes.

	Resource counts:
		Old: 2
		New: 2

	Catalogs contain the same resources by resource title


	Individual Resource differences:
	Old Resource:
	  file{"/tmp/foo":
		content => d3b07384d113edec49eaa6238ad5ff00
	  }

	New Resource:
	  file{"/tmp/foo":
		content => dbb53f3699703c028483658773628452
	  }

Had resources imply gone missing - not the case here - you would have seen
a list of that.

This code only validates the catalogs, it cannot tell you if the behavior of
the providers that interpret the catalog has changed so testing is still
recommended, this is just one tool to take away some of the uncertainty

You can get some inline help with:
  puppet man catalog

Changelog:
 - 2012/10/20 - Make it a puppet face

 - 2010/10/19 - Change options and logic so it's easier to use comparing catalogs
                from the same version of puppet

Contact:

R.I.Pienaar <rip@devco.net>
www.devco.net
@ripienaar


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org).

## [v2.2.1](https://github.com/camptocamp/puppet-catalog-diff/tree/v2.2.1) (2020-07-15)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v2.2.0...v2.2.1)

### Fixed

- Get rid of obsolete URI.escape [\#29](https://github.com/camptocamp/puppet-catalog-diff/pull/29) ([raphink](https://github.com/raphink))
- Fix problem where a node's details wouldn't load/show in puppet-catalog-diff-viewer [\#28](https://github.com/camptocamp/puppet-catalog-diff/pull/28) ([JohnEricson](https://github.com/JohnEricson))

## [v2.2.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v2.2.0) (2020-06-19)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v2.1.1...v2.2.0)

### Added

- Add --exclude\_resource\_types= flag [\#27](https://github.com/camptocamp/puppet-catalog-diff/pull/27) ([raphink](https://github.com/raphink))

## [v2.1.1](https://github.com/camptocamp/puppet-catalog-diff/tree/v2.1.1) (2020-06-10)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v2.1.0...v2.1.1)

## [v2.1.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v2.1.0) (2020-06-10)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v2.0.0...v2.1.0)

### Added

- Use \#notice instead of \#err for realtime message [\#26](https://github.com/camptocamp/puppet-catalog-diff/pull/26) ([raphink](https://github.com/raphink))
- Add --exclude\_defined\_resources [\#25](https://github.com/camptocamp/puppet-catalog-diff/pull/25) ([raphink](https://github.com/raphink))

## [v2.0.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v2.0.0) (2020-05-05)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v1.7.0...v2.0.0)

### Changed

- Decommission old fact\_search methods [\#17](https://github.com/camptocamp/puppet-catalog-diff/pull/17) ([raphink](https://github.com/raphink))

### Added

- Add coveralls [\#20](https://github.com/camptocamp/puppet-catalog-diff/pull/20) ([raphink](https://github.com/raphink))
- Test comparer [\#18](https://github.com/camptocamp/puppet-catalog-diff/pull/18) ([raphink](https://github.com/raphink))
- Update Readme [\#16](https://github.com/camptocamp/puppet-catalog-diff/pull/16) ([raphink](https://github.com/raphink))
- Refactor/lint [\#15](https://github.com/camptocamp/puppet-catalog-diff/pull/15) ([raphink](https://github.com/raphink))

### Fixed

- Fix convert\_pdb [\#21](https://github.com/camptocamp/puppet-catalog-diff/pull/21) ([raphink](https://github.com/raphink))

## [v1.7.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v1.7.0) (2020-05-05)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v1.6.0...v1.7.0)

### Added

- Refactor [\#14](https://github.com/camptocamp/puppet-catalog-diff/pull/14) ([raphink](https://github.com/raphink))
- Lint [\#13](https://github.com/camptocamp/puppet-catalog-diff/pull/13) ([raphink](https://github.com/raphink))
- Ignore parameters [\#12](https://github.com/camptocamp/puppet-catalog-diff/pull/12) ([raphink](https://github.com/raphink))
- Sort hash keys in formater [\#11](https://github.com/camptocamp/puppet-catalog-diff/pull/11) ([raphink](https://github.com/raphink))
- Include environment in report [\#10](https://github.com/camptocamp/puppet-catalog-diff/pull/10) ([raphink](https://github.com/raphink))
- Update to current PDK template [\#9](https://github.com/camptocamp/puppet-catalog-diff/pull/9) ([DavidS](https://github.com/DavidS))
- Cache str diffs [\#8](https://github.com/camptocamp/puppet-catalog-diff/pull/8) ([raphink](https://github.com/raphink))
- Puppetdb catalog [\#7](https://github.com/camptocamp/puppet-catalog-diff/pull/7) ([raphink](https://github.com/raphink))
- Output endpoint in debug [\#6](https://github.com/camptocamp/puppet-catalog-diff/pull/6) ([raphink](https://github.com/raphink))
- Support certless API [\#3](https://github.com/camptocamp/puppet-catalog-diff/pull/3) ([raphink](https://github.com/raphink))
- Use v3 catalog API [\#2](https://github.com/camptocamp/puppet-catalog-diff/pull/2) ([raphink](https://github.com/raphink))

### Fixed

- Fix: v4/catalog returns an embedded catalog [\#5](https://github.com/camptocamp/puppet-catalog-diff/pull/5) ([raphink](https://github.com/raphink))
- Check key in parsed catalog [\#1](https://github.com/camptocamp/puppet-catalog-diff/pull/1) ([raphink](https://github.com/raphink))

## [v1.6.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v1.6.0) (2015-06-04)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v1.5.1...v1.6.0)

## [v1.5.1](https://github.com/camptocamp/puppet-catalog-diff/tree/v1.5.1) (2015-06-02)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v1.5.0...v1.5.1)

## [v1.5.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v1.5.0) (2015-06-02)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/v1.0.0...v1.5.0)

## [v1.0.0](https://github.com/camptocamp/puppet-catalog-diff/tree/v1.0.0) (2014-04-04)

[Full Changelog](https://github.com/camptocamp/puppet-catalog-diff/compare/96e91f27adc96eb37ae63810b174f041add8784d...v1.0.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


diffcatalogs.rb
#!/usr/bin/env ruby

# A tool to compare catalogs that have been generated via puppet master --compile
#
# Contact:
# R.I.Pienaar <rip@devco.net> - www.devco.net - @ripienaar

require 'puppet'
require 'yaml'
require 'facter'
require 'pp'
require 'fileutils'
require 'digest/md5'

# helper methods
require 'preprocessor'
require 'comparer'

################# main program starts here #######################

Puppet.parse_config
if Puppet.version =~ /^([0-9]+[.][0-9]+)[.][0-9]+/
    @version = $1

    unless ["0.24", "0.25", "2.6", "2.7", "3.0"].include?(@version)
        puts("Don't know how to compare catalogs for version #{Puppet.version}. Only 0.24, 0.25, 2.6, and 2.7 are supported")
        exit 1
    end
else
    puts("Could not figure out version from #{Puppet.version}")
    exit 1
end

if ARGV.size == 2
    FROM = ARGV[0]
    TO = ARGV[1]
else
    puts "Please specify two catalogs to compare"
    exit 1
end

from = []
to   = []
{ FROM => from, TO => to}.each do |r,v|
    unless File.exist?(r)
        puts "Cannot find resources in #{r}"
        exit 1
    end
    
    case File.extname(r)
    when '.yaml'
      tmp = YAML.load(File.read(r))
    when '.marshal'
      tmp = Marshal.load(File.read(r))
    when '.pson'
      tmp = PSON.load(File.read(r))
    end
    
    if @version == "0.24"
      convert24(tmp, v)
    else
      convert25(tmp, v)
    end
end

titles = {}
titles[:to] = extract_titles(to)
titles[:from] = extract_titles(from)

puts "Resource counts:"
puts "\tOld: #{titles[:from].size}"
puts "\tNew: #{titles[:to].size}"
puts

puts "Resource title diffs:"
print_resource_diffs(titles[:to], titles[:from])
puts

compare_resources(from, to)


